ibis-server/tests/routers/ibis/test_bigquery.py

```diff
@@ -50,22 +53,63 @@ def get_connection_info():
             "credentials": os.getenv("TEST_BIG_QUERY_CREDENTIALS_BASE64_JSON")
         }
 
-    def test_query(self, manifest_str: str):
+    def test_query(self):
         connection_info = self.get_connection_info()
         response = client.post(
             url="/v2/ibis/bigquery/query",
             json={
                 "connectionInfo": connection_info,
-                "manifestStr": manifest_str,
-                "sql": 'SELECT * FROM "Orders" LIMIT 1'
+                "manifestStr": self.manifest_str,
+                "sql": 'SELECT * FROM "Orders" ORDER BY orderkey LIMIT 1'
+            }
+        )
+        assert response.status_code == 200
+        result = response.json()
+        assert len(result['columns']) == len(self.manifest['models'][0]['columns'])
+        assert len(result['data']) == 1
+        assert result['data'][0] == [1, 370, 'O', 172799.49, 820540800000, '1_370', 1704153599000, 1704153599000]
+        assert result['dtypes'] == {
+            'orderkey': 'int64',
+            'custkey': 'int64',
+            'orderstatus': 'object',
+            'totalprice': 'float64',
+            'orderdate': 'object',
+            'order_cust_key': 'object',
+            'timestamp': 'datetime64[ns]',
+            'timestamptz': 'datetime64[ns, UTC]'
+        }
+
+    def test_query_with_column_dtypes(self):
+        connection_info = self.get_connection_info()
+        response = client.post(
+            url="/v2/ibis/bigquery/query",
+            json={
+                "connectionInfo": connection_info,
+                "manifestStr": self.manifest_str,
+                "sql": 'SELECT * FROM "Orders" ORDER BY orderkey LIMIT 1',
+                "columnDtypes": {
+                    "totalprice": "float",
+                    "orderdate": "datetime64",
+                    "timestamp": "datetime64",
+                    "timestamptz": "datetime64"
+                }
             }
         )
         assert response.status_code == 200
         result = response.json()
-        assert len(result['columns']) == 2
+        assert len(result['columns']) == len(self.manifest['models'][0]['columns'])
         assert len(result['data']) == 1
-        assert result['data'][0][0] is not None
-        assert result['dtypes'] is not None
+        assert result['data'][0] == [1, 370, 'O', 172799.49, '1996-01-02 00:00:00.000000', '1_370', '2024-01-01 23:59:59.000000', '2024-01-01 23:59:59.000000 UTC']
+        assert result['dtypes'] == {
+            'orderkey': 'int64',
+            'custkey': 'int64',
+            'orderstatus': 'object',
+            'totalprice': 'float64',
+            'orderdate': 'object',
+            'order_cust_key': 'object',
+            'timestamp': 'object',
+            'timestamptz': 'object'
```

Comment:
Curiously, why are those object after setting the specific column type?

Reply:
Because we format the datetime to string. String in the dtype is objcet. The data type is mapping the data value. If we want to keep the type is datetime, we should not format the value. But we need the value format.

Reply:
I see. Thanks. It's good for now.